File: src/kernel/event-log.ts
// ─── Geological Event Log ───────────────────────────────────────────────────
// Records significant geological events (eruptions, collisions, rifts) with
// timestamps for display in the UI timeline and for snapshot replay.

import type { GeoEventType } from '../shared/types';

// ─── Types ──────────────────────────────────────────────────────────────────

export interface GeoLogEntry {
  /** Simulation time in Ma when the event occurred. */
  timeMa: number;
  /** Event type identifier. */
  type: GeoEventType;
  /** Human-readable description. */
  description: string;
  /** Optional cell index or lat/lon where the event occurred. */
  location?: { lat: number; lon: number };
}

// ─── EventLog ───────────────────────────────────────────────────────────────

export class EventLog {
  private entries: GeoLogEntry[] = [];

  /** Maximum stored entries to prevent unbounded growth. */
  private readonly maxEntries: number;

  constructor(maxEntries = 10_000) {
    this.maxEntries = maxEntries;
  }

  /** Record a new event. */
  record(entry: GeoLogEntry): void {
    this.entries.push(entry);
    // Trim oldest if over budget
    if (this.entries.length > this.maxEntries) {
      this.entries.splice(0, this.entries.length - this.maxEntries);
    }
  }

  /** Get all entries. */
  getAll(): ReadonlyArray<GeoLogEntry> {
    return this.entries;
  }

  /** Get entries within a time range [startMa, endMa]. */
  getRange(startMa: number, endMa: number): GeoLogEntry[] {
    return this.entries.filter(
      (e) => e.timeMa >= startMa && e.timeMa <= endMa,
    );
  }

  /** Get entries of a specific event type. */
  getByType(type: GeoEventType): GeoLogEntry[] {
    return this.entries.filter((e) => e.type === type);
  }

  /** Get the N most recent entries. */
  getRecent(count: number): GeoLogEntry[] {
    return this.entries.slice(-count);
  }

  /** Total number of recorded events. */
  get length(): number {
    return this.entries.length;
  }

  /** Clear all entries. */
  clear(): void {
    this.entries = [];
  }
}

File: backend/GeoTime.Core/Kernel/EventLog.cs
using GeoTime.Core.Models;

namespace GeoTime.Core.Kernel;

/// <summary>
/// Records significant geological events with timestamps.
/// </summary>
public sealed class EventLog(int maxEntries = 10_000)
{
    private readonly List<GeoLogEntry> _entries = [];

    public void Record(GeoLogEntry entry)
    {
        _entries.Add(entry);
        if (_entries.Count > maxEntries)
            _entries.RemoveRange(0, _entries.Count - maxEntries);
    }

    public IReadOnlyList<GeoLogEntry> GetAll() => _entries;

    public List<GeoLogEntry> GetRange(double startMa, double endMa)
        => _entries.Where(e => e.TimeMa >= startMa && e.TimeMa <= endMa).ToList();

    public List<GeoLogEntry> GetByType(string type)
        => _entries.Where(e => e.Type == type).ToList();

    public List<GeoLogEntry> GetRecent(int count)
        => _entries.Skip(Math.Max(0, _entries.Count - count)).ToList();

    public int Length => _entries.Count;

    public void Clear() => _entries.Clear();
}
